lib/rubocop/cop/lint/redundant_with_object.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # This cop checks for redundant `with_object`.
+      #
+      # @example
+      #   # bad
+      #   ary.each_with_object([]) do |v|
+      #     v
+      #   end
+      #
+      #   # good
+      #   ary.each do |v|
+      #     v
+      #   end
+      #
+      #   # bad
+      #   ary.each.with_object([]) do |v|
+      #     v
+      #   end
+      #
+      #   # good
+      #   ary.each do |v|
+      #     v
+      #   end
+      #
+      class RedundantWithObject < Cop
+        MSG_EACH_WITH_OBJECT = 'Use `each` instead of `each_with_object`.'
+                               .freeze
+        MSG_WITH_OBJECT = 'Remove redundant `with_object`.'.freeze
+
+        def_node_matcher :redundant_with_object?, <<-PATTERN
```

Comment:
I was expecting this pattern to be the same as the one of `with_index` with `index` swapped for `object`, but this pattern is slightly different. The pattern for `with_object` has an extra `(_)` after `with_object` and before `args`. Is it needed for this pattern? Should it be added to the `with_index` pattern?

```ruby
def_node_matcher :redundant_with_index?, <<-PATTERN
  (block
    $(send
      _ {:each_with_index :with_index})
    (args
      (arg _))
    ...)
PATTERN
```

Reply:
Yes. As you pointed out, the `with_index` pattern of `Lint/RedundantWithObject` cop can be improved.